### ISSUE TITLE
Quarkus Awt extension doesnt work after upgrading to 3.2.3-Final when using native build

### ISSUE THREAD

**Issue by sreekanth-pricer** [quarkusio/quarkus#35256]
### Describe the bug

We use Font class from java-awt to parse the font. After upgrading the quarkus from 3.0.3 to 3.2.3-Final. The feature is broken and i get the error "no awt in java.library.path" with the exception ""java.lang.UnsatisfiedLinkError". 

### Expected behavior

The expected is to not to get the exceptions and should be able to use Font class from java.awt

### Actual behavior

 i get the error "no awt in java.library.path" with the exception ""java.lang.UnsatisfiedLinkError".

**Comment by geoand** [comment]
cc @Karm

**Comment by gsmet** [comment]
Is it when starting your app from command line or in a Docker container?

**Comment by sreekanth-pricer** [comment]
It is failing in the docker container

**Comment by gsmet** [comment]
Yeah, so that's unfortunately expected. GraalVM changed how .so dependencies are handled.

See how it was handled here: https://github.com/quarkusio/quarkus-quickstarts/commit/a6c7cfce1c0df574592a53b41783588480c0e811#diff-89914cf05ea1b37167e374cb1625de38c8f5d48823d4ad228b15b0a3b34ca9e7R30-R32 .

I don't see it in the migration guide, though. I'll have a look later.

**Comment by sreekanth-pricer** [comment]
I still get the same error even if i do the same changes as mentioned in https://github.com/quarkusio/quarkus-quickstarts/commit/a6c7cfce1c0df574592a53b41783588480c0e811#diff-89914cf05ea1b37167e374cb1625de38c8f5d48823d4ad228b15b0a3b34ca9e7R30-R32 .

**Comment by Karm** [comment]
> I still get the same error even if i do the same changes as mentioned in [quarkusio/quarkus-quickstarts@a6c7cfc#diff-89914cf05ea1b37167e374cb1625de38c8f5d48823d4ad228b15b0a3b34ca9e7R30-R32](https://github.com/quarkusio/quarkus-quickstarts/commit/a6c7cfce1c0df574592a53b41783588480c0e811#diff-89914cf05ea1b37167e374cb1625de38c8f5d48823d4ad228b15b0a3b34ca9e7R30-R32) .

The message just means that the chain of native .so dependencies that was supposed to be loaded in order for awt to work failed.
It does a poor job at stating exactly what failed, which .so was missing or which .so failed to load .e.g  due to a missing symbol etc.

Can you share exactly which builder image or GraalVM/Mandrel distribution you use?

**Comment by sreekanth-pricer** [comment]
Hi, Please find the details below ( extracted from our pipeline)

Java version: 17.0.8+7, vendor version: Mandrel-23.0.1.2-Final

Builder Image : 
Pulling from quarkus/ubi-quarkus-mandrel-builder-image
Digest: sha256:bfb896cd4e6fff84c77bceeb722dd3e07d23f9e18709b01e09af0d44e9feb0f6
Status: Image is up to date for quay.io/quarkus/ubi-quarkus-mandrel-builder-image:jdk-17
quay.io/quarkus/ubi-quarkus-mandrel-builder-image:jdk-17

**Comment by Karm** [comment]
@sreekanth-pricer , so, how does your app differ from this:

```
$ git clone https://github.com/quarkusio/quarkus-quickstarts.git
$ cd quarkus-quickstarts/awt-graphics-rest-quickstart/
$ ./mvnw package -Pnative -Dquarkus.native.container-build=true -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel-builder-image:jdk-17
$ docker build -f src/main/docker/Dockerfile.native -t quarkus/awt-graphics-rest .
$ docker run -i --rm -p 8080:8080 quarkus/awt-graphics-rest
```

?

I invite you specifically to check the `freetype` and `fontconfig` *runtime* image dependencies in the Dockerfile. 

There is still a chance that your app does something that requires a lib I don't have covered in tests. In that case, would it be possible for you to fork and modify the quickstart app to include your Font parsing code as a quick reproducer?

**Comment by sreekanth-pricer** [comment]
I have a doubt which docker file was supposed to be used, because the Dockerfile.native-micro contains the libfreetype dependency which we use. The other docker Dockerfile.native doesn't have any of these. Could you please explain how it works.

**Comment by Karm** [comment]
@sreekanth-pricer 

* Dockerfile.native - installs dependencies using microdnf as you can see if you take a look https://github.com/quarkusio/quarkus-quickstarts/blob/main/awt-graphics-rest-quickstart/src/main/docker/Dockerfile.native#L24

* Dockerfile.native-micro is targeted to create even smaller image, does not have a capacity for running package manager, so it copies needed files over from another image.


If you are doing this yourself, using your own base and runtime images, note the BEWARE warning:
https://github.com/quarkusio/quarkus-quickstarts/blob/main/awt-graphics-rest-quickstart/src/main/docker/Dockerfile.native-micro#L7

e.g. it won't work if you are trying to copy libraries from newer glibc based Ubuntu Latest to an older glibc based RHEL 8 etc. 

It would really make things much faster if you could share more about your setup.  e.g. "We have this Linux image X, it's roughly based on XYZ etc."

**Comment by Karm** [comment]
NOTE: The freetype and fontconfig dependency is **not new**, so if it worked for you before the Quarkus upgrade, it should keep working fine.  The only news is the suite of .so files that must go with your quarkus native executable now.

**Comment by sreekanth-pricer** [comment]
We have used jib to build the native image. So I have created a base image for jib as below. The same has been used for both the quarkus versions 3.0 and 3.2 . Worked for 3.0 but not 3.2

FROM quay.io/quarkus/quarkus-micro-image:2.0
COPY --from=BUILD \
   /lib64/libfreetype.so.6 \
   /lib64/libgcc_s.so.1 \
   /lib64/libbz2.so.1 \
   /lib64/libpng16.so.16 \
   /lib64/libm.so.6 \
   /lib64/libbz2.so.1 \
   /lib64/libexpat.so.1 \
   /lib64/libuuid.so.1 \
   /lib64/

COPY --from=BUILD \
   /usr/lib64/libfontconfig.so.1 \
   /usr/lib64/

COPY --from=BUILD \
    /usr/share/fonts /usr/share/fonts

COPY --from=BUILD \
    /usr/share/fontconfig /usr/share/fontconfig

COPY --from=BUILD \
    /usr/lib/fontconfig /usr/lib/fontconfig

COPY --from=BUILD \
     /etc/fonts /etc/fonts

ENV LD_LIBRARY_PATH='/usr/lib64:/lib64'

**Comment by Karm** [comment]
@sreekanth-pricer Hmmm, when you inspect the final runtime image, can you see your quarkus application executable and its necessary .so files along with it? I mean e.g. , in my case:

```
$ docker run -it --entrypoint /bin/ls quarkus/awt-graphics-rest-quickstart /work/ -1
application
libawt.so
libawt_headless.so
libawt_xawt.so
libfontmanager.so
libjava.so
libjavajpeg.so
libjvm.so
liblcms.so
libmlib_image.so
```

**Comment by Karm** [comment]
@sreekanth-pricer ?

**Comment by sreekanth-pricer** [comment]
Hi Karm, I had all the necessary files for my font parsing as i have used a separate base image for the jib  as mentioned here https://github.com/quarkusio/quarkus/issues/35256#issuecomment-1678741643 . Also, we have moved out this font parsing from our quarkus application as we had to fix asap.

**Comment by Karm** [comment]
@gefloh Could you whip up a small reproducer project I can use with my AWS account? AWS Lambda is a Linux running somewhere and we must copy the needed libs alongside the main Quarkus application executable to it.

**Comment by gefloh** [comment]
Ok, I will fiddle something.

**Comment by gefloh** [comment]
@Karm I created a small reproducer [here](https://gitlab.pachamamita.de/pub/reproducer/quarkus-awt-native-lambda-problem) which tries to convert some HTML into a PDF. When you compile this natively, upload the function.zip to AWS and run the lambda, you get the above mentioned UnsatisfiedLinkError.

**Comment by AlexCraddock** [comment]
I have managed to solve the bug for our case that @billapepper was asking about.

In our case we had this in our docker native build in a file called .dockerignore

    *
    !target/*-runner
    !target/*-runner.jar
    !target/lib/*
    !target/quarkus-app/*

When the native build runs it's adds native libraries to the builds /target directory.

When the Dockerfile.native was building the copy command

`COPY target/*.so /work`

Was ignoring the *.so files due to the * in .dockerignore.

I have updated the .dockerignore file to 

    *
    !target/*.so
    !target/*-runner
    !target/*-runner.jar
    !target/lib/*
    !target/quarkus-app/*



Now the native libraries are included in our docker image and the AWT processing for image resizing works.



As a node a lot of the examples in https://github.com/quarkusio/quarkus-quickstarts/tree/main for example
https://github.com/quarkusio/quarkus-quickstarts/blob/main/spring-web-quickstart/.dockerignore have the same incorrect setup, as a lot of people possibly copy these examples as a starting point, might be worth checking there to make sure that the *.so files are actually being included in your docker image under /work.

**Comment by Karm** [comment]
@AlexCraddock That is an excellent catch.
I will try to propagate this throughout the ecosystem.

**Comment by Karm** [comment]
@gefloh Thx. I'll take a look with my AWS.

**Comment by Karm** [comment]
> @Karm I created a small reproducer [here](https://gitlab.pachamamita.de/pub/reproducer/quarkus-awt-native-lambda-problem) which tries to convert some HTML into a PDF. When you compile this natively, upload the function.zip to AWS and run the lambda, you get the above mentioned UnsatisfiedLinkError.

Hello @gefloh, the problem is indeed just the missing libraries.
I used your code to create my own AWS Lambda and it fails the way you described:

## :x: 

![Screenshot 2023-09-04 at 12-59-04 awttest - Lambda](https://github.com/quarkusio/quarkus/assets/691097/d3edba07-7759-4552-967f-b4db0d0f454b)

## :heavy_check_mark: 

When I manually added the .so files and re-uploaded the Lambda, it all went well:

![OK-Screenshot 2023-09-04 at 13-09-49 awttest - Lambda](https://github.com/quarkusio/quarkus/assets/691097/993b65e2-b09a-42bc-b345-9f80935ac1cc)

...returning it me a bunch base64 encoded bytes. When I decoded it and jammed it into a PDF file, it open just fine for me, showing a caption "Test":

![screenshot](https://github.com/quarkusio/quarkus/assets/691097/5802403b-b5e8-448e-b1e7-bfc81628e451)

I'll proceed and fix this in Quarkus presently.

**Comment by gefloh** [comment]
Hello @Karm , ah perfect, thanks for your support.

**Comment by rsvoboda** [comment]
PR #35718 is merged, does it also fix this issue or just https://github.com/quarkusio/quarkus/issues/35713?

**Comment by Karm** [comment]
Closing. Users need to copy libs and install libs as necessary as indicted above.

**Comment by aylwyne** [comment]
@Karm I have a question on how you solved this for an AWS Lambda function.  You said that you "manually added the .so files and re-uploaded the Lambda, it all went well".  Is there a way to get this to happen as part of the quarkus build?  I have a build that runs as a github action that uses the quarkus maven command to build the function zip file then uses terraform to apply this to AWS.  I'm not sure I see how I would get those .so files included in the function.zip.

**Comment by gefloh** [comment]
@aylwyne Just use a Quarkus version >=3.4.0. The build adds the necessary files automatically to the function.zip.

**Comment by aylwyne** [comment]
> @aylwyne Just use a Quarkus version >=3.4.0. The build adds the necessary files automatically to the function.zip.

Thanks.  I somehow missed the comment a few back linking to the specific fix in 3.4.0.
